rustls/src/server/handy.rs

```diff
@@ -166,36 +172,14 @@ impl ResolvesServerCertUsingSni {
         // These checks are not security-sensitive.  They are the
         // *server* attempting to detect accidental misconfiguration.
 
-        // Always reject an empty certificate chain.
-        let end_entity_cert = ck.end_entity_cert().map_err(|_| {
-            Error::General("No end-entity certificate in certificate chain".to_string())
-        })?;
-
-        // Reject syntactically-invalid end-entity certificates.
-        let end_entity_cert =
-            webpki::EndEntityCert::try_from(end_entity_cert.as_ref()).map_err(|_| {
-                Error::General(
-                    "End-entity certificate in certificate chain is syntactically invalid"
-                        .to_string(),
-                )
-            })?;
-
-        // Note that this doesn't fully validate that the certificate is valid; it only validates that the name is one
-        // that the certificate is valid for, if the certificate is
-        // valid.
-        let general_error =
-            || Error::General("The server certificate is not valid for the given name".to_string());
-
-        let name = webpki::DnsNameRef::try_from_ascii(checked_name.as_ref().as_bytes())
-            .map_err(|_| general_error())?;
-
-        end_entity_cert
-            .verify_is_valid_for_subject_name(webpki::SubjectNameRef::DnsName(name))
-            .map_err(|_| general_error())?;
-
-        let as_str: &str = checked_name.as_ref();
-        self.by_name
-            .insert(as_str.to_string(), Arc::new(ck));
+        ck.end_entity_cert()
+            .and_then(ParsedCertificate::try_from)
+            .and_then(|cert| verify_server_name(&cert, &server_name))?;
```

Comment:
I think this refactoring has the effect of not rejecting an empty certificate chain, or a malformed end entity cert. Were you thinking the value of that early err return isn't worthwhile?

I know this isn't security sensitive, but in the absence of a counter argument it feels like we could keep the early misconfiguration detections while still relying on `verify_server_name` for the heavy lifting.

Reply:
Are you sure you understand `and_then()` correctly?

Reply:
I think this is ok -- in this commit you can see the tests for those cases changing to track the different errors being produced -- so a difference in behaviour but the error checking is still happening.

- empty cert chain: `Error::General` -> `Error::NoCertificatesPresented`
- malformed cert: `Error::General` -> `Error::InvalidCertificate(CertificateError::BadEncoding)`
- name doesn't appear in cert: `Error::General` -> `Error::InvalidCertificate(CertificateError::NotValidForName)`

Reply:
:sweat_smile: My bad.

I was confused and thinking about `Option.and_then` yielding `None`.

Nevermind me :bow: